tests/test_stochatreat.py

```diff
@@ -277,6 +277,135 @@ def test_stochatreat_only_misfits(probs):
+    and returns the difference of the two"""
+    counts = get_within_strata_counts(treats)
+
+    required_props = pd.DataFrame({"required_prop": probs, "treat": range(len(probs))})
+    
+    comparison_df = pd.merge(
+        counts, required_props, on="treat", how="left"
+    )
+
+    comparison_df["desired_counts"] = comparison_df["block_counts"] * comparison_df["required_prop"]
+
+    return counts["treatment_counts"] - comparison_df["desired_counts"]
+
+
+@pytest.mark.parametrize("n_treats", [2, 3, 4, 5, 10])
+@pytest.mark.parametrize(
+    "block_cols", [["dummy"], ["block1"], ["block1", "block2"]]
+)
+def test_stochatreat_within_stratas_no_probs(n_treats, block_cols, df):
+    """Test that within strata treatment assignment proportions are no more than one unit away from the required proportions with equal treatment assignment probabilities"""
+    treats = stochatreat(
+        data=df, block_cols=block_cols, treats=n_treats, idx_col="id", random_state=42
+    )
+    counts_diff = compute_counts_diff(treats, n_treats * [1 / n_treats])
+
+    np.testing.assert_equal(
+        (np.absolute(counts_diff) > 1).sum(),
+        0
+    )
+
+
+@pytest.mark.parametrize("probs", standard_probs)
+@pytest.mark.parametrize(
+    "block_cols", [["dummy"], ["block1"], ["block1", "block2"]]
+)
+def test_stochatreat_within_stratas_probs(probs, block_cols, df):
+    """Test that within strata treatment assignment proportions are no more than one unit away from the required proportions with unequal treatment assignment probabilities"""
+    treats = stochatreat(
+        data=df,
+        block_cols=block_cols,
+        treats=len(probs),
+        idx_col="id",
+        probs=probs,
+        random_state=42,
+    )
+    counts_diff = compute_counts_diff(treats, probs)
+
+    np.testing.assert_equal(
+        (np.absolute(counts_diff) > 1).sum(),
+        0
+    )
+
+
+@pytest.mark.parametrize("probs", [[0.1, 0.9], [0.5, 0.5], [0.9, 0.1]])
+def test_stochatreat_within_strata_no_misfits(probs):
+    """Test that within strata treatment assignment proportions are no more than one unit away from the required proportions when strata are such that there are no misfits"""
+    N = 10_000
+    blocksize = 10
+    df = pd.DataFrame(
+        data={
+            "id": np.arange(N),
+            "block": np.repeat(
+                np.arange(N / blocksize),
+                repeats=blocksize
+            )
+        }
+    )
+
+    treats = stochatreat(
+        data=df,
+        block_cols=['block'],
+        treats=len(probs),
+        idx_col="id",
+        probs=probs,
+        random_state=42,
+    )
+    counts_diff = compute_counts_diff(treats, probs)
+
+    np.testing.assert_equal(
+        (np.absolute(counts_diff) > 1).sum(),
+        0
+    )
+
+
+@pytest.mark.parametrize("probs", standard_probs)
+def test_stochatreat_within_strata_only_misfits(probs):
```

Comment:
depending on how you assign the misfits this need not or maybe even should not be true, right?

Reply:
Yes you're right should rather test if the sum of counts diff within each block is always less than the number of treatments (i.e len(probs) for example), right?